### DISCUSSION TITLE
How to adjust volume of a Godot/GMC bus - in this case the music bus

### DISCUSSION THREAD

**Discussion by bernarma** [missionpinball/missionpinball#147] [Q&A] [answered]
**Use Case**
I have a background song that plays throughout a player's ball. When the ball ends, the bonus mode will play (for an unknown amount of time). During that time, I would like the music to continue to play (which it currently does) but I would like it to be at a reduced volume.

**Setup**
Using MPF 0.80 (latest pre-release version) and MPF-GMC (latest Godot).

**Background**
I have tried ducking with a silent sound, loop set to -1 (infinite) and a target duck bus set to "music" with an attenuation of different values, although the loop is ignored, and the duration of the sound sample is used instead. I have also not heard any difference in the sound volume during that 2 second song loop. The Godot logging does indicate that it is performing the duck with a duration and attenuation but unfortunately, I can't hear any difference.

Thanks for your assistance - I'll persevere and try to find out the issue in the meantime and will report back if I find out the cause.

**Comment by bernarma** [comment]
> **Use Case** I have a background song that plays throughout a player's ball. When the ball ends, the bonus mode will play (for an unknown amount of time). During that time, I would like the music to continue to play (which it currently does) but I would like it to be at a reduced volume.
> 
> **Setup** Using MPF 0.80 (latest pre-release version) and MPF-GMC (latest Godot).
> 
> **Background** I have tried ducking with a silent sound, loop set to -1 (infinite) and a target duck bus set to "music" with an attenuation of different values, although the loop is ignored, and the duration of the sound sample is used instead. I have also not heard any difference in the sound volume during that 2 second song loop. The Godot logging does indicate that it is performing the duck with a duration and attenuation but unfortunately, I can't hear any difference.
> 
> Thanks for your assistance - I'll persevere and try to find out the issue in the meantime and will report back if I find out the cause.

OK... Some troubleshooting on my side, it would appear as though the interpretation in MPF-GMC of the volume is inconsistent with the MPF side.

Starting with Godot, (refer here - https://docs.godotengine.org/en/stable/classes/class_audiostreamplayer.html#class-audiostreamplayer-property-volume-db), the volume_db is a decibel value.

In MPF, when setting the value to a value between 0 and 1 - this is NOT a DB therefore when it goes across to Godot it is taken as is and has no effect on the volume (from what I could tell).

If I change this line (in the GMCChannel.gd file - line 91) from

```self.volume_db = settings["volume"] if settings.get("volume")```

to

```self.volume_db = linear_to_db(settings["volume"] if settings.get("volume") else 0.5)```

it works. Keep in mind I added 0.5 as the "default" as specified by the documentation if no value is specified in the volume field.

I would like some advice before submitting a fix as there may be implications with that fix and further changes required elsewhere in that codebase (as I can see references to -80db as well. I'll link this info up to the Issue I raised in the project.

**Reply by avanwinkle** [reply]
This all makes sense. I think 1.0 is the most intuitive default value, and we can update the docs. Thanks for catching this!

**Reply by bernarma** [reply, marked as answer]
> This all makes sense. I think 1.0 is the most intuitive default value, and we can update the docs. Thanks for catching this!

Submitted change: https://github.com/missionpinball/mpf-gmc/pull/14
